---
title: "Basic Bioinformatics Overview"
author: "Ashwini Patil"
date: "December 21, 2015"
output: pdf_document
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

```{r}
library(pheatmap)
library(rjson)
library(ggplot2)
library(png)
library(grid)
library(pander)
library(stringr)
```

### Introduction
In our previous workin adult, male rats, we found that a sub-population exposed to chronic social defeat over 5-7 days exhibits behavioral vurnerability (increased passive coping, anxiety-like and depressive-like behavior) whereas the other sub-population exhibits resilience to these effects of defeat. Our overall goal is to understand the neural mechanisms that produce vulnerability and resilence in order to reduce vulnerability and promote resilence. This work has transational relevance because chronic stress is associated with post-traumatic stress disorder and depression. We have also determined that circulating microRNA profiles of rats that go on to be vulnerable are different. Thus, we as interested in biomarkers that can predict future vulnerability or resilience.

This project has three goals and purposes to use fecal samples.

Goal 1:
To determine whether vulnerable and resilient rats have different gut microbiome profiles.
Hypothesis: Rats vulnerable to stress will exhibit a different gut microbiome profile compared to resilient or non-stressed controls.

Goal 2:
To determine whether the gut microbiome profile at pre-stress timepoints is different before the rats become resilient or vurnerable.
Hypothesis: Gut microbiome profiles prior to stress may predict which individuals will be vunerable or resilient to future mess.

Goal 3:
To determine whether vulnerable rats can be made resilient by pair-housing with a resilient rats.
Hypothesis: Such pair-housing may shift the vulnerable phenotype to a resilient one due to transfer of microbiome through coprophagia.

Overall:
Goal 1 and 2 can be achieved through a single experiment with three group of rats (control, resilient, vulnerable) and fecal samples collected pre-stress and after the stress phenotype is established. Goal 3 is a bigger study requiring more controls. Wouls also serve to validate the initial findings in a seperate cohort of animals in addition to testing a novel way to promote resilience.

### Demultiplexing and quality control
Samples were sequences on Hiseq 2000 and were demultiplexed and analyzed for sequencing quality using FASTQC. Nextera-XT adapters were used while sequencing which were removed by the trimming of adapters using trimmomatic-0.33. A second quality control step using FASTQC was performed to check quality of sequences after adapter trimming.

**Figure 1: Number of read pairs in samples after demultiplexing**

Please see attached file, `quality_after.pdf`.

```{r,fig.width=7, fig.height=5, fig.align='center', echo=FALSE}
fp = file.path('summary-dnabc.json')
save_fp = gsub(pattern = ".json", replacement = ".pdf", fp)
save_fp_txt = gsub(pattern = ".json", replacement = ".txt", fp)

json_data <- fromJSON(file=fp)
data = t(data.frame(json_data$data)) # this line may need work
num_seq = data.frame(numSeq = data[,1], count = data[,1]/1000000)

num_seq$toPlot = row.names(num_seq) != 'unassigned'
num_seq$toPlot[grep('negative|vibriolambda|crypto', row.names(num_seq))] = FALSE

mean(num_seq$count[num_seq$toPlot])

write.table(data, file=save_fp_txt, quote=FALSE, col.names=FALSE, sep='\t')

ggplot(num_seq[num_seq$toPlot,], aes(x=count)) +
  geom_histogram() +
  theme_classic() +
  theme_bw() + 
  xlab("Number of read pairs in sample (millions, M)") +
  ylab("Number of samples") 
ggsave(filename=save_fp, width=7, height=5, units='in')
```

**Figure 2: Average nucleotide quality after adapter trimming and quality control**

Please see attached file, `quality_after.pdf`.

```{r,fig.width=7, fig.height=5, fig.align='center', echo=FALSE}
fp = file.path('fastqc_after_trim_quality.tsv')

quality = read.table(fp, sep='\t', header=TRUE, stringsAsFactors = FALSE, row.names = 1)

pos = gsub('X', '', colnames(quality))
pos = sapply(pos, function(x) mean(as.numeric(unlist(strsplit(x, '[.]')))))

df = data.frame(pos=pos,
                mn=colMeans(quality),
                std=apply(quality, 2, sd))

ggplot(df, aes(x=pos, y=mn)) + 
  geom_errorbar(aes(ymin=mn-std, ymax=mn+std), width=.1) +
  geom_line() +
  geom_point() +
  theme_bw() + 
  xlab('nucleotide position') +
  ylab('quality score')
ggsave(filename='quality_after.pdf', width=7, height=5, units='in')

```


Nextera-XT adapters were trimmed using trimmomatic and Table 1 shows the number of sequenced reads and number reads retrieved from both strands after adapter trimming.

**Table 1: Table showing number of reads before and after trimmming the adapters with Trimmomatic.**


```{r, echo=FALSE}
preprocess <- read.table("preprocess_summary.tsv", header = T)
preprocess$Samples <- substr(preprocess$Samples, 1, nchar(as.character(preprocess$Samples))-5)
illqc <- preprocess[,1:6]
colnames(illqc) <- c("Samples","BothKept", "Dropped", "ForwardOnly", "Input", "ReverseOnly")
pander(illqc, split.table = Inf)
```


### Human DNA filtering
Human DNA was filtered using BWA with HG38 version of human genome as reference. At the end of the filtering step the reads were divided into 2 files containd Human DNA and non-human DNA for each sample. Table 2 list out the number of human and non-human reads and the human percentage in the samples.

**Table 2: Table showing number of reads after filtering of Human reads.**
```{r, echo=FALSE}
preprocess <- read.table("preprocess_summary.tsv", header = T)
preprocess$Samples <- substr(preprocess$Samples, 1, nchar(as.character(preprocess$Samples))-5)
decontam <- cbind.data.frame(Samples=preprocess$Samples, false=preprocess$false, true=preprocess$true)
decontam$Human_percent <- (decontam$true * 100)/(decontam$true + decontam$false)
colnames(decontam) <- c("Samples","Non-HumanReads", "HumanReads", "HumanReadsPercent" )
pander(decontam, split.table = Inf)
```

### Taxonomic assignments using Metaphlan2
Taxonomic assignments were performned using Metaphlan2. A heatmap was generated from the results of taxonomic assignments.

**Figure 3: Heatmap generated after taxonomic assignments using Metaphlan2**

Please see attached file, `taxonomic_assignments.pdf`.

```{r,  echo=FALSE}
s <- read.delim("taxonomic_assignments.tsv", comment.char="#")
# Convert assignment column to string
s$Term <- as.character(s$Term)

# Keep only species-level rows
# We should do this in python before table is saved
is_species <- str_count(s$Term, fixed("|")) == 6
s <- s[is_species,]

# Store as a matrix
smat <- as.matrix(s[,-1])
# Check that column sums add to 1
#colSums((smat))

# Grab species names from assignment strings
species_labels <- sapply(strsplit(s$Term, "|", fixed = T), `[`, 7)
species_labels <- sub("s__", "", species_labels)
rownames(smat) <- species_labels

# Use the pheatmap library for plot
pheatmap(
  smat, filename = "taxonomic_assignments.pdf",
  cellwidth = 9, cellheight = 9, fontsize = 10)
```

### Pathway abundance 
Presence/abscence and abundance of genes in pathways was evaluated by mapping reads to KEGG database using RAPsearch and then abundace were calculated. 

**Figure 4: Heatmap generated after mapping against KEGG database showing presence/absence and abundance of genes in pathways**

Please see attached file, `gene_function_assignments.pdf`.

```{r,echo=FALSE}
s <- read.delim("ko_assignments.tsv", comment.char="#")
rownames(s) <- s$Term
s$Term <- NULL
pheatmap(
  s, filename = "gene_function_assignments.pdf",
  cellwidth = 9, cellheight = 9, fontsize = 10)
```